presto-orc/src/main/java/com/facebook/presto/orc/reader/LongDictionaryStreamReader.java

```diff
@@ -123,66 +120,63 @@ public Block readBlock(Type type)
-            if (presentStream == null) {
-                if (dataStream == null) {
-                    throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+
+        if (presentStream == null) {
+            if (dataStream == null) {
+                throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+            }
+            if (inDictionaryStream == null) {
+                //dataStream.nextIntVector(nextBatchSize, idsVector, 0);
+                for (int i = 0; i < nextBatchSize; i++) {
+                    type.writeLong(builder, dictionary[((int) dataStream.next())]);
                 }
-                Arrays.fill(nullVector, false);
-                dataStream.nextLongVector(subBatchSize, dataVector);
             }
             else {
-                int nullValues = presentStream.getUnsetBits(subBatchSize, nullVector);
-                if (nullValues != subBatchSize) {
-                    if (dataStream == null) {
-                        throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+                for (int i = 0; i < nextBatchSize; i++) {
+                    long id = dataStream.next();
+                    if (inDictionaryStream.nextBit()) {
+                        type.writeLong(builder, dictionary[(int) id]);
+                    }
+                    else {
+                        type.writeLong(builder, id);
                     }
-                    dataStream.nextLongVector(subBatchSize, dataVector, nullVector);
                 }
             }
-
-            if (inDictionaryStream == null) {
-                Arrays.fill(inDictionaryVector, true);
+        }
+        else {
+            // presentStream != null
+            if (dataStream == null) {
+                // Just check if it's all null values
+                int nullValues = presentStream.getUnsetBits(nextBatchSize);
+                if (nullValues != nextBatchSize) {
+                    throw new OrcCorruptionException(streamDescriptor.getOrcDataSourceId(), "Value is not null but data stream is not present");
+                }
             }
             else {
-                inDictionaryStream.getSetBits(subBatchSize, inDictionaryVector, nullVector);
-            }
-
-            for (int i = 0; i < subBatchSize; i++) {
-                if (nullVector[i]) {
-                    builder.appendNull();
-                }
-                else if (inDictionaryVector[i]) {
-                    type.writeLong(builder, dictionary[((int) dataVector[i])]);
-                }
-                else {
-                    type.writeLong(builder, dataVector[i]);
+                for (int i = 0; i < nextBatchSize; i++) {
+                    if (!presentStream.nextBit()) {
+                        // value is null.
```

Comment:
I don't think this and the next comment are necessary.... the code is pretty straight forward